firefox-ios/Client/Frontend/UserContent/UserScripts/AllFrames/AtDocumentStart/LoginsHelper.js

```diff
@@ -578,6 +578,70 @@ window.__firefox__.includeOnce("LoginsHelper", function() {
     },
   }
 
+  // define the field types for focus events
+  const FocusFieldType = {
+    username: "username",
+    password: "password"
+  };
+
+  // Extended the LoginManagerContent object with a new method for focus events
+  LoginManagerContent._addFocusListenerBasedOnHeuristics = function (form) {
+    // common form structure
+    var commonStructures = `input[type="text"], input[type="email"], input:not([type])`
+    var possibleUsernameFields = Array.from(form.querySelectorAll(commonStructures));
+    var passwordFields = form.querySelectorAll('input[type="password"]');
+
+    // add listener to password fields
+    passwordFields.forEach((field) => this._addFocusListener(field, FocusFieldType.password));
+
+    // determine possible username field
+    possibleUsernameFields = possibleUsernameFields.filter(field =>
+      field.placeholder.toLowerCase().includes("user") ||
+      field.placeholder.toLowerCase().includes("email") ||
+      field.id.toLowerCase().includes("user") ||
+      field.id.toLowerCase().includes("email") ||
+      field.name.toLowerCase().includes("user") ||
+      field.name.toLowerCase().includes("email"));
+
+    // prioritize fields closer to the first password field, if available
+    if (passwordFields.length > 0) {
+      const firstPasswordIndex = Array.from(form.elements).indexOf(passwordFields[0]);
+      possibleUsernameFields = possibleUsernameFields.filter(field =>
+        Array.from(form.elements).indexOf(field) < firstPasswordIndex);
+    }
+
+    // attach listeners to possible username fields, if not already focused on password fields
+    possibleUsernameFields.forEach((field) => this._addFocusListener(field, FocusFieldType.username));
+
+    // no clear username field is identified & password fields are present,
+    // attach to the first text input before the password field as a fallback
+    if (possibleUsernameFields.length === 0 && passwordFields.length > 0) {
+      let precedingTextFields = Array.from(passwordFields[0].parentNode.querySelectorAll('input[type="text"], input[type="email"], input:not([type])'));
+      if (precedingTextFields.length > 0) {
+        // Last text input before the first password field
+        let assumedUsernameField = precedingTextFields[precedingTextFields.length - 1];
+        this._addFocusListener(assumedUsernameField, FocusFieldType.username);
+      }
+    }
```

Comment:
This has some edge cases that we solve already in desktop that will be tricky if we want to solve them here. Mentioning a few  just to illustrate:
- `input[type="text"], input[type="email"], input:not([type])` selector is not exhaustive. For instance a username can be a `tel` number. 
- Github for instance uses this placeholder for an input to invite collaborators `placeholder="Search by team name, username, full name, or email"` this will be identified wrongly as a username. 
- We also allow here username only forms which is a tricky case to handle and we might have a lot of edge cases with this implementation. We do [quite a few things](https://searchfox.org/mozilla-central/rev/a8cc31504a2379bcf8ba395d2da7bb632b5521d6/toolkit/components/passwordmgr/LoginManagerChild.sys.mjs#868) in desktop to make sure we don't wrongly identify a username only form.

I suggest for now, we handle only forms that have both username and passwords. Then most of the edge cases should be covered by [`inputTypeIsCompatibleWithUsername`](https://github.com/mozilla-mobile/firefox-ios/blob/b5155bed5580c781bb6005d201e3be0553bbc4b6/firefox-ios/Client/Assets/CC_Script/LoginManager.shared.mjs#L10-L16) shared from desktop. The other ones will need better heuristics that we can share later. 

Here is what I am thinking; instead of querying all forms and filtering out the inputs, we should be listening on `focusin` event and then figuring out whether we are in a form or not. Something like this should work and behave the same for the most part:

```js
  function onFocusIn(event) {
    const form = event.target?.form;
    if (!form) {
      return;
    }

    const [username, password] = LoginManagerContent._getFormFields(form, false);
    if (username && password) {
      webkit.messageHandlers.loginsManagerMessageHandler.postMessage({
        type: "fieldType",
        fieldType: event.target === username ? "username" : "password",
      });
    }
  }

  document.addEventListener("focusin", (ev) => onFocusIn(ev));
```
  
  This has a couple of advantages:
  - We don't need to select any forms on load or listen to DOM changes ( also we can now remove that `MutationObserver`). When something is focused we are notified.
  - We don't have to write custom filtering logic. `_getFormFields` takes care of that ( once we share a better heuristic from desktop it would be a drop-in replacement for this)

Reply:
Oh my, good call
When in future we are doing the work on the platform side then this does seem overkill 😅 
I tested your change locally and it works! Pushing now